stdlib/public/core/StringNormalization.swift

```diff
@@ -102,3 +102,253 @@ internal func _tryNormalize(
+  guts: _StringGuts,
+  outputBuffer: UnsafeMutableBufferPointer<UInt8>,
+  icuInputBuffer: UnsafeMutableBufferPointer<UInt16>,
+  icuOutputBuffer: UnsafeMutableBufferPointer<UInt16>
+) -> NormalizationResult {
+  _internalInvariant(guts.isFastUTF8)
+  return guts.withFastUTF8 { sourceBuffer in
+    let sourceCount = sourceBuffer.count - readIndex.encodedOffset
+    let start = readIndex.encodedOffset
+    let rebasedSourceBuffer = UnsafeBufferPointer(rebasing: sourceBuffer[start...])
+    do {
+      let (read, filled) = fastFill(rebasedSourceBuffer, outputBuffer)
+        if filled > 0 {
+          let nextIndex = readIndex.encoded(offsetBy: read)
+          _internalInvariant(guts.isOnUnicodeScalarBoundary(nextIndex))
+
+          return .success(amountFilled: filled, nextReadPosition: nextIndex)
+        }
+    }
+    guard let (read, filled) = transcodeToUTF16(rebasedSourceBuffer, into: icuInputBuffer) else {
+      return .bufferTooSmall(count: sourceBuffer.count)
+    }
+    
+    let nextIndex = readIndex.encoded(offsetBy: read)
+    _internalInvariant(guts.isOnUnicodeScalarBoundary(nextIndex))
+    
+    let rebasedICUInputBuffer = UnsafeBufferPointer(rebasing: icuInputBuffer[..<filled])
+    return sharedNormalize(
+      sourceCount, nextIndex, outputBuffer, rebasedICUInputBuffer, icuOutputBuffer
+    )
+  }
+}
+
+internal func foreignNormalize(
+  readIndex: String.Index,
+  guts: _StringGuts,
+  outputBuffer: UnsafeMutableBufferPointer<UInt8>,
+  icuInputBuffer: UnsafeMutableBufferPointer<UInt16>,
+  icuOutputBuffer: UnsafeMutableBufferPointer<UInt16>
+) -> NormalizationResult {
+  let sourceCount = guts.count - readIndex.encodedOffset
+  guard let (read, filled) = foreignFill(readIndex, guts, into: icuInputBuffer) else {
+    return .bufferTooSmall(count: guts.count)
+  }
+  
+  let nextIndex = readIndex.encoded(offsetBy: read)
+  _internalInvariant(guts.isOnUnicodeScalarBoundary(nextIndex))
+  
+  let rebasedICUInputBuffer = UnsafeBufferPointer(rebasing: icuInputBuffer[..<filled])
+  return sharedNormalize(
+    sourceCount, nextIndex, outputBuffer, rebasedICUInputBuffer, icuOutputBuffer
+  )
+}
+
+func foreignFill(
+  _ readIndex: String.Index,
+  _ guts: _StringGuts,
+  into outputBuffer: UnsafeMutableBufferPointer<UInt16>
+) -> (Int, Int)? {
+  var index = readIndex
+  var writeIndex = 0
+  let outputCount = outputBuffer.count
+  let cachedEndIndex = guts.endIndex
+  while index != cachedEndIndex {
+    let (scalar, length) = guts.foreignErrorCorrectedScalar(startingAt: index)
+    if scalar._hasNormalizationBoundaryBefore && index != readIndex {
+      break
+    }
+    
+    index = index.encoded(offsetBy: length)
+    
+    for cu in scalar.utf16 {
+      if writeIndex < outputCount {
+        outputBuffer[writeIndex] = cu
+        writeIndex += 1
+      } else {
+        return nil
+      }
+    }
+  }
+  return (index.encodedOffset - readIndex.encodedOffset, writeIndex)
+}
+
+private func sharedNormalize(
```

Comment:
Shared meaning shared strings, or shared code, or what? Please add comments to these methods. If this is a core common implementation method, `Impl` or is probably better than `Shared`